File: src/mcp/tools/definitions/event_create.ts
/**
 * event_createツール定義
 * タイムラインにイベントを追加するMCPツール
 */

import type {
  McpToolDefinition,
  ToolExecutionContext,
} from "../tool_registry.ts";
import { executeCliCommand } from "../cli_adapter.ts";
import { ElementEventCommand } from "../../../cli/modules/element/event.ts";

/**
 * 有効なイベントカテゴリ
 */
const validCategories = [
  "plot_point",
  "character_event",
  "world_event",
  "backstory",
  "foreshadow",
  "climax",
  "resolution",
];

export const eventCreateTool: McpToolDefinition = {
  name: "event_create",
  description:
    "既存のタイムラインにイベントを追加します。物語の出来事をタイムライン上に配置し、因果関係を記録できます。",
  inputSchema: {
    type: "object",
    properties: {
      timelineId: {
        type: "string",
        description: "イベントを追加するタイムラインのID",
      },
      title: {
        type: "string",
        description: "イベントのタイトル",
      },
      category: {
        type: "string",
        enum: validCategories,
        description:
          "イベントのカテゴリ: plot_point（プロットポイント）, character_event（キャラクターイベント）, world_event（世界イベント）, backstory（バックストーリー）, foreshadow（伏線）, climax（クライマックス）, resolution（解決）",
      },
      order: {
        type: "number",
        description: "タイムライン上の順序（数値）",
      },
      summary: {
        type: "string",
        description: "イベントの概要",
      },
      id: {
        type: "string",
        description: "イベントID（省略時はtitleから自動生成）",
      },
      characters: {
        type: "array",
        items: { type: "string" },
        description: "関連するキャラクターIDのリスト",
      },
      settings: {
        type: "array",
        items: { type: "string" },
        description: "関連する設定（場所等）IDのリスト",
      },
      chapters: {
        type: "array",
        items: { type: "string" },
        description: "関連するチャプターIDのリスト",
      },
      causedBy: {
        type: "array",
        items: { type: "string" },
        description: "このイベントの原因となるイベントIDのリスト",
      },
      causes: {
        type: "array",
        items: { type: "string" },
        description: "このイベントが引き起こすイベントIDのリスト",
      },
      importance: {
        type: "string",
        enum: ["major", "minor", "background"],
        description: "イベントの重要度",
      },
    },
    required: ["timelineId", "title", "category", "order"],
  },
  execute: async (
    args: Record<string, unknown>,
    context?: ToolExecutionContext,
  ) => {
    const timelineId = args.timelineId as string | undefined;
    const title = args.title as string | undefined;
    const category = args.category as string | undefined;
    const order = args.order as number | undefined;

    // 必須パラメータのチェック
    if (
      !timelineId || typeof timelineId !== "string" || timelineId.trim() === ""
    ) {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'timelineId' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    if (!title || typeof title !== "string" || title.trim() === "") {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'title' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    if (!category || typeof category !== "string") {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'category' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    if (!validCategories.includes(category)) {
      return {
        content: [
          {
            type: "text" as const,
            text: `Error: Invalid category '${category}'. Must be one of: ${
              validCategories.join(", ")
            }.`,
          },
        ],
        isError: true,
      };
    }

    if (order === undefined || typeof order !== "number") {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'order' parameter is required and must be a number.",
          },
        ],
        isError: true,
      };
    }

    // CLIコマンド用の引数を構築（CLIの引数名にマッピング）
    const commandArgs: Record<string, unknown> = {
      timeline: timelineId, // MCP: timelineId → CLI: timeline
      title,
      category,
      order,
    };

    if (typeof args.summary === "string") {
      commandArgs.summary = args.summary;
    }

    if (typeof args.id === "string") {
      commandArgs.id = args.id;
    }

    // 配列パラメータをカンマ区切り文字列に変換
    if (Array.isArray(args.characters)) {
      const characters = (args.characters as unknown[])
        .filter((c) => typeof c === "string")
        .map((c) => (c as string).trim())
        .filter((c) => c.length > 0);
      if (characters.length > 0) {
        commandArgs.characters = characters.join(",");
      }
    } else if (typeof args.characters === "string") {
      commandArgs.characters = args.characters;
    }

    if (Array.isArray(args.settings)) {
      const settings = (args.settings as unknown[])
        .filter((s) => typeof s === "string")
        .map((s) => (s as string).trim())
        .filter((s) => s.length > 0);
      if (settings.length > 0) {
        commandArgs.settings = settings.join(",");
      }
    } else if (typeof args.settings === "string") {
      commandArgs.settings = args.settings;
    }

    if (Array.isArray(args.chapters)) {
      const chapters = (args.chapters as unknown[])
        .filter((c) => typeof c === "string")
        .map((c) => (c as string).trim())
        .filter((c) => c.length > 0);
      if (chapters.length > 0) {
        commandArgs.chapters = chapters.join(",");
      }
    } else if (typeof args.chapters === "string") {
      commandArgs.chapters = args.chapters;
    }

    if (Array.isArray(args.causedBy)) {
      const causedBy = (args.causedBy as unknown[])
        .filter((c) => typeof c === "string")
        .map((c) => (c as string).trim())
        .filter((c) => c.length > 0);
      if (causedBy.length > 0) {
        commandArgs["caused-by"] = causedBy.join(",");
      }
    } else if (typeof args.causedBy === "string") {
      commandArgs["caused-by"] = args.causedBy;
    }

    if (Array.isArray(args.causes)) {
      const causes = (args.causes as unknown[])
        .filter((c) => typeof c === "string")
        .map((c) => (c as string).trim())
        .filter((c) => c.length > 0);
      if (causes.length > 0) {
        commandArgs.causes = causes.join(",");
      }
    } else if (typeof args.causes === "string") {
      commandArgs.causes = args.causes;
    }

    if (typeof args.importance === "string") {
      commandArgs.importance = args.importance;
    }

    const handler = new ElementEventCommand();
    return await executeCliCommand(handler, commandArgs, context?.projectRoot);
  },
};

File: internal/mcp/tools/event_create.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/takets/street-storyteller/internal/domain"
	"github.com/takets/street-storyteller/internal/mcp/protocol"
)

// EventCreateTool creates a new TimelineEvent in the given timeline.
type EventCreateTool struct{}

type eventCreateArgs struct {
	TimelineID string   `json:"timeline_id"`
	Title      string   `json:"title"`
	Category   string   `json:"category"`
	Summary    string   `json:"summary"`
	Order      int      `json:"order"`
	Characters []string `json:"characters"`
	Settings   []string `json:"settings"`
	Chapters   []string `json:"chapters"`
}

// Definition advertises the event_create schema.
func (EventCreateTool) Definition() protocol.Tool {
	return protocol.Tool{
		Name:        "event_create",
		Description: "Create a new event in a timeline",
		InputSchema: json.RawMessage(`{
			"type": "object",
			"required": ["timeline_id", "title", "category", "summary"],
			"properties": {
				"timeline_id": {"type": "string", "description": "Target timeline ID"},
				"title":       {"type": "string", "description": "Event title"},
				"category":    {"type": "string", "description": "Event category (plot_point, character_event, world_event, backstory, foreshadow, climax, resolution)"},
				"summary":     {"type": "string", "description": "Event summary"},
				"order":       {"type": "integer", "description": "Relative order within timeline (default 0)"},
				"characters":  {"type": "array", "items": {"type": "string"}, "description": "Character IDs involved"},
				"settings":    {"type": "array", "items": {"type": "string"}, "description": "Setting IDs involved"},
				"chapters":    {"type": "array", "items": {"type": "string"}, "description": "Chapter IDs related"}
			}
		}`),
	}
}

// Handle validates args and returns a constructed TimelineEvent as JSON.
func (EventCreateTool) Handle(_ context.Context, args json.RawMessage, _ ExecutionContext) (*protocol.CallToolResult, error) {
	var a eventCreateArgs
	if len(args) > 0 {
		_ = json.Unmarshal(args, &a)
	}

	if a.TimelineID == "" {
		return errorResult(fmt.Errorf("timeline_id is required")), nil
	}
	if a.Title == "" {
		return errorResult(fmt.Errorf("title is required")), nil
	}
	if a.Category == "" {
		return errorResult(fmt.Errorf("category is required")), nil
	}
	if a.Summary == "" {
		return errorResult(fmt.Errorf("summary is required")), nil
	}

	// Generate event ID from order position.
	// Why: simple deterministic ID rather than UUID — keeps output predictable
	// and avoids importing extra dependencies for a tool that doesn't persist data.
	eventID := fmt.Sprintf("event_%d", a.Order)

	chars := a.Characters
	if chars == nil {
		chars = []string{}
	}
	settings := a.Settings
	if settings == nil {
		settings = []string{}
	}
	chapters := a.Chapters
	if chapters == nil {
		chapters = []string{}
	}

	ev := domain.TimelineEvent{
		ID:       eventID,
		Title:    a.Title,
		Category: domain.EventCategory(a.Category),
		Time:     domain.TimePoint{Order: a.Order},
		Summary:  a.Summary,
		Characters: chars,
		Settings:   settings,
		Chapters:   chapters,
	}

	evJSON, err := json.MarshalIndent(ev, "", "  ")
	if err != nil {
		return errorResult(fmt.Errorf("marshal event: %w", err)), nil
	}

	text := fmt.Sprintf("event created: %s in %s\n%s", eventID, a.TimelineID, string(evJSON))
	return &protocol.CallToolResult{
		Content: []protocol.ContentBlock{{Type: "text", Text: text}},
	}, nil
}
